crypto/awskms/signer.go

```diff
@@ -0,0 +1,149 @@
+func (c *Client) SignerForKey(
+	ctx context.Context,
+	key Key,
+) (*Signer, error) {
+	pk, hashAlgo, err := c.GetPublicKey(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Signer{
+		ctx:       ctx,
+		client:    c.client,
+		key:       key,
+		curve:     pk.Algorithm(),
+		publicKey: pk,
+		hashAlgo:  hashAlgo,
+	}, nil
+}
+
+// Sign signs the given message using the KMS signing key for this signer.
+//
+// Reference: https://github.com/aws/aws-sdk-go-v2/blob/main/service/kms/api_op_Sign.go
+func (s *Signer) Sign(message []byte) ([]byte, error) {
+
+	keyArn := s.key.ARN()
+	// AWS KMS supports signing messages without pre-hashing
+	// up to 4096 bytes. Beyond that limit, messages must be prehashed outside KMS.
+	kmsPreHashLimit := 4096
+	var request *kms.SignInput
+
+	if len(message) <= kmsPreHashLimit {
+		request = &kms.SignInput{
+			KeyId:            &keyArn,
+			Message:          message,
+			SigningAlgorithm: types.SigningAlgorithmSpecEcdsaSha256,
+		}
+	} else {
+		// this is guaranteed to only return supported hash algos by KMS
+		hasher, err := crypto.NewHasher(s.hashAlgo)
+		if err != nil {
+			return nil, fmt.Errorf("awskms: failed to sign: %w", err)
+		}
+		// pre-hash outside KMS
+		hash := hasher.ComputeHash(message)
+		// indicate the MessageType is digest
+		request = &kms.SignInput{
+			KeyId:            &keyArn,
+			Message:          hash,
+			SigningAlgorithm: types.SigningAlgorithmSpecEcdsaSha256,
+			MessageType:      types.MessageTypeDigest,
+		}
+	}
+	result, err := s.client.Sign(s.ctx, request)
+	if err != nil {
+		return nil, fmt.Errorf("awskms: failed to sign: %w", err)
+	}
+	sig, err := parseSignature(result.Signature, s.curve)
+	if err != nil {
+		return nil, fmt.Errorf("awskms: failed to parse signature: %w", err)
+	}
+	return sig, nil
+}
+
+// parseSignature parses an asn1 stucture (R,S) into a slice of bytes as required by the `Siger.Sign` method.
+func parseSignature(kmsSignature []byte, curve crypto.SignatureAlgorithm) ([]byte, error) {
+	var parsedSig struct{ R, S *big.Int }
+	if _, err := asn1.Unmarshal(kmsSignature, &parsedSig); err != nil {
+		return nil, fmt.Errorf("asn1.Unmarshal: %w", err)
+	}
+
+	curveOrderLen := curveOrder(curve)
+	signature := make([]byte, 2*curveOrderLen)
+
+	// left pad R and S with zeroes
+	rBytes := parsedSig.R.Bytes()
+	sBytes := parsedSig.S.Bytes()
+	copy(signature[curveOrderLen-len(rBytes):], rBytes)
+	copy(signature[len(signature)-len(sBytes):], sBytes)
+
+	return signature, nil
+}
+
+// returns the curve order size in bytes (used to padd R and S of the ECDSA signature)
+// Only P-256 and secp256k1 are supported. The calling function should make sure
+// the function is only called with one of the 2 curves.
+func curveOrder(curve crypto.SignatureAlgorithm) int {
+	switch curve {
+	case crypto.ECDSA_P256:
+		return 32
+	case crypto.ECDSA_secp256k1:
+		return 32
+	default:
+		panic("the curve is not supported")
+	}
+}
```

Comment:
Thanks for adding `crypto/kms_internal`. I think if the the sub-package is called `internal`, the capital methods (here `ParseSignature `) won't be exported outside `crypto/`.  But I'm not sure if that remains true if the package only has "internal" suffix. Can you confirm that's the case? otherwise we could just use "internal"